### PR TITLE
Add DeletionProtection mechanism reject Namespace deletion when PVCs are included under NS

### LINKED ISSUE #1223: [feature request] The DeletionProtection mechanism reject Namespace deletion when PVCs are included under NS
**What would you like to be added**:
When there is active Pods in this namespace, openkruise can reject Namespace Deletion, details: https://openkruise.io/docs/user-manuals/deletionprotection .

But pvc is also a very important resource, so if pvc exists under Namespace, NS is also not allowed to delete.
Related Code: https://github.com/openkruise/kruise/blob/master/pkg/webhook/util/deletionprotection/deletion_protection.go#L52

### PR DESCRIPTION
**Pull request by kevin1689-cloud**
### Ⅰ. Describe what this PR does
This PR complete a feature: The DeletionProtection mechanism reject Namespace deletion when PVCs are included under NameSpace

### Ⅱ. Does this pull request fix one issue?

fixes #1223 

### Ⅲ. Describe how to verify it
1.Create a namespace and label it with "policy.kruise.io/delete-protection=Cascading"
2.Create a pvc in that namespace
3.Try to delete the namespce, the namespace can't be deleted and show an error which tell user the reason is there are existing pvc in the namespce

### Ⅳ. Special notes for reviews
None

### CODE REVIEW

**Comment by zmberg** [comment]
@kevin1689-cloud good, can you can some E2E in [code](https://github.com/openkruise/kruise/blob/master/test/e2e/policy/deletionprotection.go#L113)?

**Comment by kevin1689-cloud** [comment]
> @kevin1689-cloud good, can you can some E2E in [code](https://github.com/openkruise/kruise/blob/master/test/e2e/policy/deletionprotection.go#L113)?

Sure, l will working on that.

**Comment by zmberg** [comment]
@furykerry Is only PVC bound to PV worth protecting?

**Comment by zmberg** [comment]
@kevin1689-cloud I think PVC which unbound to pv or deleting are not worth protecting.

**Comment by kevin1689-cloud** [comment]
> @kevin1689-cloud I think PVC which unbound to pv or deleting are not worth protecting.

@zmberg  Got it, as the [Deletion Protection](https://openkruise.io/docs/user-manuals/deletionprotection/) mechanism is aim to enhance applications' availability, I agree with the PVC which not bound to PV is not worth protecting. I will modify my code and E2E to achieve that. 

I checked [kubernetes code](https://github.com/kubernetes/api/blob/e3f1032953d3776ef5da09db363f00babf681791/core/v1/types.go#L673), PVC have 3 status: Pending, Bound, Lost. So the logic is when PVC's status is in "Bound", protect the NameSpace to be delete, and when PVC is in other status, the NameSpace can be delete, is that correct?

**Comment by zmberg** [comment]
> > @kevin1689-cloud I think PVC which unbound to pv or deleting are not worth protecting.
> 
> @zmberg Got it, as the [Deletion Protection](https://openkruise.io/docs/user-manuals/deletionprotection/) mechanism is aim to enhance applications' availability, I agree with the PVC which not bound to PV is not worth protecting. I will modify my code and E2E to achieve that.
> 
> I checked [kubernetes code](https://github.com/kubernetes/api/blob/e3f1032953d3776ef5da09db363f00babf681791/core/v1/types.go#L673), PVC have 3 status: Pending, Bound, Lost. So the logic is when PVC's status is in "Bound", protect the NameSpace to be delete, and when PVC is in other status, the NameSpace can be delete, is that correct?

Yes.

**Comment by kevin1689-cloud** [comment]
> > > @kevin1689-cloud I think PVC which unbound to pv or deleting are not worth protecting.
> > 
> > 
> > @zmberg Got it, as the [Deletion Protection](https://openkruise.io/docs/user-manuals/deletionprotection/) mechanism is aim to enhance applications' availability, I agree with the PVC which not bound to PV is not worth protecting. I will modify my code and E2E to achieve that.
> > I checked [kubernetes code](https://github.com/kubernetes/api/blob/e3f1032953d3776ef5da09db363f00babf681791/core/v1/types.go#L673), PVC have 3 status: Pending, Bound, Lost. So the logic is when PVC's status is in "Bound", protect the NameSpace to be delete, and when PVC is in other status, the NameSpace can be delete, is that correct?
> 
> Yes.

Ok.

**Comment by kevin1689-cloud** [comment]
@zmberg Done. Now namespace will be protected only when PVCs in "Bound" status are included under the namespce. The relevant E2E tests have been added. Please take a look.

File: pkg/webhook/util/deletionprotection/deletion_protection.go [hunk 1]
```diff
@@ -71,6 +71,21 @@ func ValidateNamespaceDeletion(c client.Client, namespace *v1.Namespace) error {
...
 		}
+
+		pvcs := v1.PersistentVolumeClaimList{}
+		if err := c.List(context.TODO(), &pvcs, client.InNamespace(namespace.Name)); err != nil {
+			return fmt.Errorf("forbidden by ResourcesProtectionDeletion for list pvc error: %v", err)
+		}
+		var boundCount int
+		for i := range pvcs.Items {
+			pvc := &pvcs.Items[i]
+			if pvc.Status.Phase == v1.ClaimBound {
```
Comment (zmberg) [inline pkg/webhook/util/deletionprotection/deletion_protection.go:82]: if pvc.deletionTimestamp == nil && pvc.Status.Phase == v1.ClaimBound {

}

File: test/e2e/policy/deletionprotection.go [hunk 2]
```diff
@@ -19,6 +19,7 @@ package policy
 import (
 	"context"
 	"fmt"
+	"k8s.io/apimachinery/pkg/api/resource"
```
Comment (zmberg) [inline test/e2e/policy/deletionprotection.go:22]: go imports the 'import' section.

**Comment by kevin1689-cloud** [comment]
@zmberg Done. I made a misoperation when uploading the code and submitted several recent updates from others. Now I have fixed this misoperation. If there are any problem, please tell me. Thanks!

File: pkg/webhook/util/deletionprotection/deletion_protection.go [hunk 3]
```diff
@@ -71,6 +71,21 @@ func ValidateNamespaceDeletion(c client.Client, namespace *v1.Namespace) error {
 		if activeCount > 0 {
 			return fmt.Errorf("forbidden by ResourcesProtectionDeletion for %s=%s and active pods %d>0", policyv1alpha1.DeletionProtectionKey, val, activeCount)
 		}
+
+		pvcs := v1.PersistentVolumeClaimList{}
+		if err := c.List(context.TODO(), &pvcs, client.InNamespace(namespace.Name)); err != nil {
```
Comment (furykerry) [inline pkg/webhook/util/deletionprotection/deletion_protection.go:76]: plz add  DisableDeepCopy option here

Comment (zmberg) [reply pkg/webhook/util/deletionprotection/deletion_protection.go:76]: and add DisableDeepCopy option in list pods function again.

Comment (kevin1689-cloud) [reply pkg/webhook/util/deletionprotection/deletion_protection.go]: @furykerry @zmberg Done. Now the list pvc and list pods function is as shown below:
```
pvcs := v1.PersistentVolumeClaimList{}
if err := c.List(context.TODO(), &pvcs, client.InNamespace(namespace.Name), utilclient.DisableDeepCopy);
```
```
pods := v1.PodList{}
if err := c.List(context.TODO(), &pods, client.InNamespace(namespace.Name), utilclient.DisableDeepCopy);
```
Please take a look.

**Review by furykerry — COMMENTED** [review]
/lgtm

**Pull request merged**